### PR TITLE
[Core] Support image processor

### LINKED ISSUE #4054: [Bug]: Incorrect Data Type Conversion for MultiModalData
### Your current environment

```text
The output of `python collect_env.py`
```

### 🐛 Describe the bug

I've encountered an issue with the code in the vllm project on GitHub. Specifically, at line 170 of the llm.py file https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/llm.py#L170, there is a conversion of multi_modal_data.data to torch.float16 using the statement multi_modal_data.data = multi_modal_data.data.to(torch.float16). This automatic conversion may not be suitable for all use cases, especially if the model is designed to operate with bfloat16 or other numerical precisions. If this is an issue that needs to be addressed, I would be happy to submit a pull request with a fix

### PR DESCRIPTION
**Pull request by DarkLight1337**
I have implemented a plugin architecture (`MultiModelPlugin`) over `MultiModalData` to define how each modality type should be preprocessed before being passed to the model as keyword arguments. This preserves the contract between the output of HuggingFace processor and the input into the HuggingFace model. As long as those keyword arguments do not conflict with the ones we have in vLLM, I think this is a good way to make the framework flexible enough to support other multi-modal architectures.

FIX #4054 (the data is now automatically converted into the model's device)

### Related Contributions

This PR is part of #3978.

This PR also implements Proposals 1 and 3 of #4194.

### Features
1. Refactor of `MultiModalData`
    - Image input is now split into two subclasses:
        - `ImageFeatureData` represents the image features of LLaVA after being passed through the vision tower, but before the multi-modal projection is applied.
        - `ImagePixelData` represents the raw image (using `PIL.Image` class). `AutoImageProcessor` from HuggingFace is loaded from `config.json` to pre-process input images before being passed to the model as `pixel_values`. As with the tokenizer, you can override the default one and specify the version of image processor via `EngineConfig`; you can even disable image preprocessing altogether, which is useful if you want to pass in images that have already been preprocessed.
    - The LLaVA implementation has been updated accordingly to accept the new inputs.
2. A new documentation page for using VLMs can be found under ([dev/multimodal](https://github.com/DarkLight1337/vllm-rocm/blob/mm-data-processor/docs/source/dev/multimodal.rst)).

### Compatibility Changes

- `pillow` will be upgraded to a `common` dependency (from `dev`) to process the images.

### CODE REVIEW

**Comment by DarkLight1337** [comment]
The LLaVA test passes on my end (with the output matching the HF output shown in CI). Does anyone have a clue what might cause it to fail in CI? Perhaps a case of floating-point error in GPU computation?

**Review by ywang96 — COMMENTED** [review]
Thank you for the great work and I left a few comments! I've also reviewed your other PR https://github.com/vllm-project/vllm/pull/4910, so let's get that merged first then come back to this.

P.S I'm also working on a GPT4V API server PR as a follow up once we merge this one.

File: vllm/engine/arg_utils.py
Comment (DarkLight1337) [reply vllm/engine/arg_utils.py:652]: I guess that would make it easier to quickly turn on/off image processing. I'll downgrade the error to a warning then.

File: vllm/worker/model_runner.py
Comment (DarkLight1337) [reply vllm/worker/model_runner.py:400]: It's used in the line `mm_kwargs = self.multi_modal_input_processor(mm_data)`.

Comment (ywang96) [reply vllm/worker/model_runner.py:401]: Yeah - what I meant really was that we can simply do since `mm_data` is not used anywhere else in this function:
`mm_kwargs = self.multi_modal_input_processor(seq_group_metadata.multi_modal_data)`

Comment (DarkLight1337) [reply vllm/worker/model_runner.py:401]: The variable is also used in the `if` statement. I made an intermediate variable to avoid exceeding the line limit in `mm_kwargs = self.multi_modal_input_processor(seq_group_metadata.multi_modal_data)`.

**Comment by DarkLight1337** [comment]
> Thank you for the great work and I left a few comments! I've also reviewed your other PR #4910, so let's get that merged first then come back to this.
> 
> P.S I'm also working on a GPT4V API server PR as a follow up once we merge this one.

Thanks for your time! I have updated the code according to your comments. During this process, I noticed that I didn't update the code for some model runners so I fixed them as well.

File: vllm/multimodal/registry.py
Comment (DarkLight1337) [reply vllm/multimodal/registry.py:35]: I have added a new section to the vLLM documentation to explain this.

**Comment by DarkLight1337** [comment]
Before merging this PR, imo we should complete #4328 first as it simplifies the API of passing multi-modal data.

**Comment by DarkLight1337** [comment]
I have merged #4328 into this PR in advance, so you might see additional diffs. These diffs should disappear once #4328 is merged into `main`.

**Comment by DarkLight1337** [comment]
Note that I have split out some of the changes in `conftest.py` to #5118.

**Comment by DarkLight1337** [comment]
Note that I have removed `multi_modal_data` from the legacy LLM API since this PR will break existing code that uses it anyway (due to the new method of constructing `MultiModalData`).

File: vllm/multimodal/base.py
Comment (ywang96) [inline vllm/multimodal/base.py:32]: Why do we need to define these two? Are these defined for line length purposes?

File: vllm/multimodal/image.py
Comment (ywang96) [inline vllm/multimodal/image.py:21]: Let's add a `NOTE` that this currently assumes image data is simply concatenated with prompt.

File: vllm/multimodal/image.py
Comment (ywang96) [inline vllm/multimodal/image.py:17]: Missing return type hint (same with a few other functions in this file)

Comment (ywang96) [inline vllm/multimodal/image.py:17]: Is this only for dummy data generation? If so let's rename it
```suggestion
def _get_dummy_seq_data(seq_len: int, vlm_config: VisionLanguageConfig):
```

File: vllm/multimodal/image.py
Comment (ywang96) [inline vllm/multimodal/image.py:24]:
```suggestion
def _get_dummy_values(vlm_config: VisionLanguageConfig):
```

File: vllm/multimodal/image.py
Comment (ywang96) [inline vllm/multimodal/image.py:92]: We should add a `NOTE` reference on this why this is the default way to get pixel values of an `PIL.Image.Image`

File: vllm/multimodal/registry.py
Comment (ywang96) [inline vllm/multimodal/registry.py:155]: IMO the name `MM_REGISTRY` is slightly confusing and informal since we will be using this a lot.

What's the benefit of this over making all functions of `MultiModalRegistry` static methods and directly do something like
`@MultiModalRegistry.register_image_feature_input()`

**Review by ywang96 — COMMENTED** [review]
@DarkLight1337 Left another round of reviews - PTAL, thanks!

File: vllm/multimodal/base.py
Comment (DarkLight1337) [reply vllm/multimodal/base.py:32]: They provide the generic type parameters used in `MultiModalPlugin` and `MultiModalRegistry`.

File: vllm/multimodal/registry.py
Comment (DarkLight1337) [reply vllm/multimodal/registry.py:155]: Using a constant makes it more obvious that it's a global instance. I frequently encounter this while dealing with 3D vision models (e.g. [this repo](https://github.com/search?q=repo%3APointcept%2FPointcept%20register&type=code))

Comment (DarkLight1337) [reply vllm/multimodal/registry.py:155]: If `MM_REGISTRY` is too vague, we can use the full name `MULTIMODAL_REGISTRY`. Wdyt?

Comment (ywang96) [reply vllm/multimodal/registry.py]: I see - yea this makes sense then

**Review by DarkLight1337 — COMMENTED** [review]
Forgot to remove vllm/worker/model_runner.py#L531-533, but I turned off my PC already. Can you do it for me?

File: examples/llava_example.py
Comment (ywang96) [inline examples/llava_example.py:32]: Should we add an example here that if `--disable-image-processor` is specified, then we need to actually pass in pixel values (i.e, `images/stop_sign_pixel_values.pt`)?

File: examples/llava_example.py
Comment (ywang96) [inline examples/llava_example.py:26]:
```suggestion
```

**Review by ywang96 — COMMENTED** [review]
Some final nits - once I test out the example again we should be good to go!

File: examples/llava_example.py
Comment (DarkLight1337) [reply examples/llava_example.py:32]: Sure!

**Comment by DarkLight1337** [comment]
I have enabled tensors to be passed to `ImagePixelData` to skip preprocessing. However, it seems that the pixel tensor has slightly different preprocessing than HuggingFace's `LLaVAProcessor`, causing the test to fail.

**Comment by ywang96** [comment]
Per offline discussion - waiting for #5118 to be merged first.

**Comment by ywang96** [comment]
@DarkLight1337 Could you resolve the merge conflicts? Once that's done I think this PR is ready to merge.

File: vllm/model_executor/models/llava.py
Comment (ywang96) [inline vllm/model_executor/models/llava.py:89]: Note for a later PR: we should add update doc regarding this additional registry step for adding new multi-modal language models

**Review by ywang96 — APPROVED** [review]
I did a final pass and left a note, but everything LGTM! Thank you for the hard work on this. @DarkLight1337

File: vllm/model_executor/models/llava.py
Comment (DarkLight1337) [reply vllm/model_executor/models/llava.py:89]: Will be added in #5276

**Comment by nittaya3** [comment]
Hello. I used my new profile yesterday. to add you guys And I want to see new people too. But I'm in France with you all the same.

**Pull request merged**
